File: solutions/python/easy/p0242_valid_anagram.py

"""LeetCode 242: Valid Anagram.

Problem: https://leetcode.com/problems/valid-anagram/
"""


class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        """Return whether t is an anagram of s.

        Time: O(n)
        Space: O(n)
        """
        if len(s) != len(t):
            return False

        count = {}

        for char in s:
            if char not in count:
                count[char] = 1
            else:
                count[char] += 1

        for char in t:
            if char not in count or count[char] == 0:
                return False
            count[char] -= 1
        return True


if __name__ == "__main__":
    solution = Solution()
    print(solution.isAnagram("anagram", "nagaram"))  # Expected: True
    print(solution.isAnagram("rat", "car"))          # Expected: False

File: solutions/java/easy/P0242_ValidAnagram.java
/**
 * LeetCode 242: Valid Anagram.
 *
 * Problem: https://leetcode.com/problems/valid-anagram/
 */
public class P0242_ValidAnagram {
    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length()){
            return false;
        }

        int [] counts = new int[26];

        for (int i = 0; i < s.length(); i++) {
            char letter = s.charAt(i);
            int index = letter - 'a';
            counts[index] = counts[index] + 1;
        }

        for (int i = 0; i < t.length(); i++) {
            char letter = t.charAt(i);
            int index = letter - 'a';
            counts[index] = counts[index] - 1;
        }

        for (int i = 0; i < counts.length; i++) {
            if (counts[i] != 0) {
                return false;
            }
        }

        return true;
    }

    public static void main(String[] args) {
        P0242_ValidAnagram solution = new P0242_ValidAnagram();
        System.out.println(solution.isAnagram("anagram", "nagaram")); // Expected: true
        System.out.println(solution.isAnagram("rat", "car"));         // Expected: false
    }
}
